crates/services/importer/src/importer.rs

```diff
@@ -0,0 +1,241 @@
+use crate::{
+    ports::{
+        BlockVerifier,
+        Executor,
+        ExecutorDatabase,
+        ImporterDatabase,
+    },
+    Config,
+};
+use anyhow::anyhow;
+use fuel_core_storage::{
+    transactional::StorageTransaction,
+    IsNotFound,
+};
+use fuel_core_types::{
+    blockchain::{
+        consensus::{
+            Consensus,
+            Sealed,
+        },
+        primitives::BlockHeight,
+        SealedBlock,
+    },
+    fuel_vm::crypto::ephemeral_merkle_root,
+    services::{
+        block_importer::{
+            ImportResult,
+            UncommittedResult,
+        },
+        executor::{
+            ExecutionBlock,
+            ExecutionResult,
+        },
+        Uncommitted,
+    },
+};
+use std::sync::Arc;
+use tokio::sync::broadcast;
+
+pub struct Importer<D, E, V> {
+    database: D,
+    executor: E,
+    verifier: V,
+    broadcast: broadcast::Sender<Arc<ImportResult>>,
+    guard: tokio::sync::Mutex<()>,
+}
+
+impl<D, E, V> Importer<D, E, V> {
+    pub fn new(config: Config, database: D, executor: E, verifier: V) -> Self {
+        let (broadcast, _) = broadcast::channel(config.max_block_notify_buffer);
+        Self {
+            database,
+            executor,
+            verifier,
+            broadcast,
+            guard: Default::default(),
+        }
+    }
+
+    pub fn subscribe(&self) -> broadcast::Receiver<Arc<ImportResult>> {
+        self.broadcast.subscribe()
+    }
+}
+
+impl<IDatabase, E, V> Importer<IDatabase, E, V>
+where
+    IDatabase: ImporterDatabase,
+{
+    /// The method commits the result of the block execution attaching the consensus data.
+    /// It expects that the `UncommittedResult` contains the result of the block
+    /// execution(It includes the block itself), but not more.
+    ///
+    /// It doesn't do any checks regarding block validity(execution, fields, signing, etc.).
+    /// It only checks the validity of the database.
+    ///
+    /// After the commit into the database notifies about a new imported block.
+    ///
+    /// Only one commit may be in progress at the time. All other calls will be suspended until
+    /// the first is done. It uses the `tokio::sync::Mutex`, so the parent task will wake up
+    /// when it is ready.
+    pub async fn commit_result<EDatabase>(
+        &self,
+        result: UncommittedResult<StorageTransaction<EDatabase>>,
+    ) -> anyhow::Result<()>
+    where
+        EDatabase: ExecutorDatabase,
+    {
+        let _guard = self.guard.lock().await;
```

Comment:
This is only protecting the database from being committed to. If we allow blocks to execute in any order then try to force them in order on commit then the data they are executing on will be stale. Also this creates a race condition where imported headers error because they both pass validation but then try to commit at the same height

Reply:
If you try to execute new blocks in any order, you will face an error that some data is unavailable. By locking here, I tried to highlight that we are really worrying about the "commit" part.
So it should not be possible, but you are right about stale resources, let's lock on `execute_and_commit` too.